README.md
brothermfc
==========

Access Brother MFC web UI details with node.js

[![npm](https://img.shields.io/npm/v/brothermfc.svg?maxAge=3600)](https://github.com/fvdm/nodejs-brothermfc/blob/master/CHANGELOG.md)
[![Build Status](https://travis-ci.org/fvdm/nodejs-brothermfc.svg?branch=master)](https://travis-ci.org/fvdm/nodejs-brothermfc)
[![Dependency Status](https://gemnasium.com/badges/github.com/fvdm/nodejs-brothermfc.svg)](https://gemnasium.com/github.com/fvdm/nodejs-brothermfc#runtime-dependencies)


Example
-------

```js
var brother = require ('brothermfc') ({
  hostname: 'myprinter.lan'
});

brother.current (function (err, data) {
  console.log (err || data);
});
```


Installation
------------

Normal: `npm install brothermfc`

Development: `npm install fvdm/nodejs-brothermfc`


Configuration
-------------

param    | type   | required | default | description
:--------|:-------|:---------|:--------|:------------------------------------------
hostname | string | yes      |         | Printer hostname or IP
port     | number | no       | 80      | Printer web UI port
protocol | string | no       | http    | Printer web UI protocol, `http` or `https`
prefix   | string | no       |         | Printer web UI prefix
ippPort  | number | no       | 631     | Printer IPP service port
timeout  | number | no       | 5000    | Web require time out in ms, 1000 = 1 sec


#### Example

```js
var brother = require ('brothermfc') ({
  hostname: 'printer.lan',
  protocol: 'https'
});
```


current
-------
**( callback )**

Get current information, like display text and ink levels, directly
from the IPP interface. This is the most accurate information compared
to the `general.status` method which is a calculation.

**Note:** this method may also work with other models and brands.


param    | type     | required | description
:--------|:---------|:---------|:-----------
callback | function | yes      | see Usage.


#### Example

```js
brother.current (callback);
```

#### Output

```js
{ state: 'idle',
  stateReasons: 'none',
  jobs: 0,
  uptime: 613845,
  uptimeDate: Tue Nov 18 2014 03:47:22 GMT+0100 (CET),
  ink: {
    magenta: 100,
    cyan: 100,
    yellow: 98,
    black: 86
  }
}
```


property     | description
:------------|:--------------------------------------
state        | short status message
stateReasons | more descriptive status message
jobs         | print tasks running and queued
uptime       | seconds since last boot
uptimeDate   | `Date` object from `uptime`
ink          | inklevels in %, colors depend in model


general.status
--------------
**( callback )**

Get basic status details, like display text and ink levels.

**Note:** `ink` levels are less accurate then the `.current` method.


param    | type     | required | description
:--------|:---------|:---------|:-----------
callback | function | yes      | see Usage.


#### Example

```js
brother.general.status (callback);
```


#### Output

```js
{ model: 'MFC-J4710DW',
  status: 'moniOk',
  message: 'Gereed',
  ink: { magenta: 100, cyan: 100, yellow: 100, black: 91 },
  contact: 'Franklin',
  location: 'Here' }
```


general.information
-------------------
**( callback )**

Get information specific to the device, i.e. IP-address and serialnumber.

param    | type     | required | description
:--------|:---------|:---------|:-----------
callback | function | yes      | see Usage.

```js
brother.general.information (callback);
```


#### Output

```js
{ node_name: 'BRA123BC456DE78',
  model_name: 'Brother MFC-J4710DW',
  location: 'Here',
  contact: 'Franklin',
  ip_address: '192.168.1.158',
  serial_no: 'A12345B678',
  firmware_version: 'H',
  page_counter: 26 }
```


general.sleep
-------------
**( [preset], callback )**

Get or change the device's sleep timeout.

param    | type     | required | description
:--------|:---------|:---------|:-----------
preset   | number   | no       | Preset number corresponding to minutes
callback | function | yes      | see [Example](#example)


### Get

```js
brother.general.sleep (callback);
```


#### Output

```js
{
  presets: {
    1: 2,
    2: 3,
    3: 5,
    4: 10,
    5: 30,
    6: 60
  ],
  value: {
    key: 4, minutes: 10
  }
}
```

The `presets` property are the predefined minutes you can choose from.
The `value` property is the current setting.


### Change

```js
// set to key `5` = 30 minutes
brother.general.sleep (5, callback);
```


#### Output

* OK: `data` is `true` (boolean)
* Fail: `err` is `Error: POST failed` with `err.error`


License
-------

This is free and unencumbered software released into the public domain.

Anyone is free to copy, modify, publish, use, compile, sell, or
distribute this software, either in source code form or as a compiled
binary, for any purpose, commercial or non-commercial, and by any
means.

In jurisdictions that recognize copyright laws, the author or authors
of this software dedicate any and all copyright interest in the
software to the public domain. We make this dedication for the benefit
of the public at large and to the detriment of our heirs and
successors. We intend this dedication to be an overt act of
relinquishment in perpetuity of all present and future rights to this
software under copyright law.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
OTHER DEALINGS IN THE SOFTWARE.

For more information, please refer to <http://unlicense.org>


Author
------

[Franklin van de Meent](https://frankl.in)


CHANGELOG.md
#### 1.0.2 (2016-5-26)

##### Documentation Changes

* **readme:** Add version badge for changelog ([a6de067f](https://github.com/fvdm/nodejs-brothermfc/commit/a6de067f1171e4d1e319e998b55f8148735184ac))

##### Other Changes

* **undefined:**
  * add node v6 to Travis config ([64add4a1](https://github.com/fvdm/nodejs-brothermfc/commit/64add4a10a5b8df511852738b37a7614f65af3ab))
  * added dependencies badge ([b5a26959](https://github.com/fvdm/nodejs-brothermfc/commit/b5a269593f9c3abb5984a2b7e5c32027c8211083))

##### Refactors

* **package:** Include CHANGELOG.md in package ([03270d59](https://github.com/fvdm/nodejs-brothermfc/commit/03270d591def44d95c3976c81fc770df9acf03a6))



test.js
var dotest = require ('dotest');
var app = require ('./');

var config = {
  protocol: process.env.MFC_PROTOCOL || 'http',
  hostname: process.env.MFC_HOSTNAME || null,
  port: process.env.MFC_PORT || 80,
  prefix: process.env.MFC_PREFIX || '',
  ippPort: process.env.MFC_IPPPORT || 631,
  timeout: process.env.MFC_TIMEOUT || 5000
};

var cacheSleep = null;

var mfc = app (config);


dotest.add ('Module', function (test) {
  var general = mfc && mfc.general;

  test ()
    .isFunction ('fail', 'exports', app)
    .isObject ('fail', 'interface', mfc)
    .isFunction ('fail', '.current', mfc && mfc.current)
    .isFunction ('fail', '.sleep', mfc && mfc.sleep)
    .isObject ('fail', '.general', general)
    .isFunction ('fail', '.general.status', general && general.status)
    .isFunction ('fail', '.general.information', general && general.information)
    .done ();
});


dotest.add ('Configuration', function (test) {
  if (!config.hostname) {
    test ()
      .warn ('MFC_HOSTNAME not set, skipping method tests')
      .exit ();
  } else {
    test ()
      .good ('MFC_HOSTNAME is set')
      .info ('MFC_TIMEOUT:  ' + config.timeout)
      .done ();
  }
});


dotest.add ('Method .sleep - get value', function (test) {
  mfc.sleep (function (err, data) {
    var value = data && data.value;

    if (value && value.key) {
      cacheSleep = value.key;
    }

    test (err)
      .isObject ('fail', 'data', data)
      .isObject ('fail', 'data.presets', data && data.presets)
      .isObject ('fail', 'data.value', value)
      .isNumber ('fail', 'data.value.key', value && value.key)
      .isNumber ('fail', 'data.presets[1]', data && data.presets && data.presets[1])
      .done ();
  });
});


dotest.add ('Method .sleep - set value', function (test) {
  mfc.sleep (cacheSleep, function (err, data) {
    test (err)
      .isExactly ('fail', 'data', data, true)
      .done ();
  });
});


dotest.add ('Method .current', function (test) {
  mfc.current (function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isString ('fail', 'data.state', data && data.state)
      .isString ('fail', 'data.stateReasons', data && data.stateReasons)
      .isNumber ('fail', 'data.jobs', data && data.jobs)
      .isNumber ('fail', 'data.uptime', data && data.uptime)
      .isDate ('fail', 'data.uptimeDate', data && data.uptimeDate)
      .isObject ('fail', 'data.ink', data && data.ink)
      .info (data && data.ink)
      .done ();
  });
});


dotest.add ('Method general.status', function (test) {
  mfc.general.status (function (err, data) {
    var i;

    test (err)
      .isObject ('fail', 'data', data)
      .isString ('fail', 'data.model', data && data.model)
      .isString ('fail', 'data.status', data && data.status)
      .isString ('fail', 'data.message', data && data.message)
      .isObject ('fail', 'data.ink', data && data.ink)
      .isNotEmpty ('fail', 'data.ink', data && data.ink);

    for (i in data && data.ink) {
      test ()
        .isNumber ('fail', 'data.ink.' + i, data.ink [i]);
    }

    test ()
      .done ();
  });
});


dotest.add ('Method general.information', function (test) {
  mfc.general.information (function (err, data) {
    test (err)
      .isObject ('fail', 'data', data)
      .isString ('fail', 'data.ip_address', data && data.ip_address)
      .isNumber ('fail', 'data.page_counter', data && data.page_counter)
      .done ();
  });
});


dotest.run ();


brothermfc.js
/*
Name:         brothermfc
Description:  Access Brother MFC web UI details with node.js
Author:       Franklin van de Meent (https://frankl.in)
Feedback:     https://github.com/fvdm/nodejs-brothermfc/issues
Source:       https://github.com/fvdm/nodejs-brothermfc
License:      Unlicense / Public Domain (see LICENSE file)
*/

var ipp = require ('ipp');
var httpreq = require ('httpreq');

var config = {
  timeout: 5000,
  protocol: 'http',
  hostname: null,
  port: 80,
  prefix: '',
  ippPort: 631
};


/**
 * Process response
 *
 * @callback callback
 * @param response {object} - httpreq response details
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function doResponse (response, callback) {
  var data = response && response.body || '';
  var error = null;

  var res = {
    code: response.statusCode,
    headers: response.headers,
    data: data.trim ()
  };

  if (response.statusCode >= 300) {
    error = new Error ('HTTP error');
    error.code = res.code;
    callback (error);
    return;
  }

  callback (null, res);
}


/**
 * Communication
 *
 * @callback callback
 * @param props {object} - Request properties
 * @param props.path - Request path
 * @param [props.method = GET] {string} - HTTP method
 * @param [props.protocol = http] {string} - `https` or `http`
 * @param [props.hostname = config.hostname] {string} - Printer hostname
 * @param [props.port = 80] {number} - Printer web UI port
 * @param [props.headers] {object} - Additional request headers
 * @returns {void}
 */

function talk (props, callback) {
  var options = {
    method: props.method || 'GET',
    parameters: props.query || null,
    timeout: config.timeout,
    headers: {}
  };

  var port = props.protocol === 'https' ? 443 : 80;

  options.url = props.protocol || config.protocol;
  options.url += '://';
  options.url += props.hostname || config.hostname;
  options.url += ':';
  options.url += props.port || config.port || port;
  options.url += config.prefix + props.path;

  // custom request headers
  if (props.headers) {
    options.headers = props.headers;
  }

  // request
  httpreq.doRequest (options, function (err, res) {
    if (err) {
      callback (err);
      return;
    }

    doResponse (res, callback);
  });
}


/**
 * Method: general.status
 *
 * @callback callback
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function methodGeneralStatus (callback) {
  talk ({ path: '/general/status.html' }, function (err, res) {
    var result = {
      model: null,
      status: null,
      message: null,
      ink: {}
    };

    if (err) {
      callback (err);
      return;
    }

    res.data.replace (/<div id="modelName"><h1>([^<]+)<\/h1>/, function (s, t) {
      result.model = t;
    });

    res.data.replace (/<div id="moni_data"><span class="moni ([^\"]+)">([^<]+)<\/span><\/div>/, function (s, t, m) {
      result.status = t;
      result.message = m;
    });

    res.data.replace (/<img src="\.\.\/common\/images\/(\w+)\.gif" alt="([^\"]+)" class="tonerremain" height="(\d+)px"/g, function (s, c, s2, p) {
      result.ink [c] = Math.round (p * 1.7857142857);
    });

    res.data.replace (/<li class="(contact|location)">[^<]+<span class="spacer">:<\/span>([^<]+)<\/li>/g, function (s, t, v) {
      result [t] = v;
    });

    callback (null, result);
  });
}


/**
 * Method: general.information
 *
 * @callback callback
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function methodGeneralInformation (callback) {
  talk ({ path: '/etc/mnt_info.csv' }, function (err, res) {
    var data;
    var one;
    var two;
    var result = {};
    var i;
    var val;
    var key;

    if (err) {
      callback (err);
      return;
    }

    data = res.data.replace (/\""{0}/g, '');
    data = data.split ('\n');

    one = data [0] .split (',');
    two = data [1] .split (',');
    result = {};

    for (i = 0; i < one.length; i++) {
      val = two [i] .trim ();
      key = one [i] .trim () .replace (/(.+)/, function (s, t) {
        return t.toLowerCase () .replace (' ', '_') .replace ('.', '');
      });

      if (key !== '' && val !== '') {
        result [key] = val.match (/^\d+$/) ? val * 1 : val;
      }
    }

    callback (null, result);
  });
}


/**
 * Method: sleep
 *
 * `set` values:
 * 0 = 1 min
 * 1 = 2 min
 * 2 = 3 min
 * 3 = 5 min
 * 4 = 10 min
 * 5 = 30 min
 * 6 = 60 min
 *
 * @callback callback
 * @param [set] {number} - 1-6, see above. Leave out to get current value.
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function methodSleep (set, callback) {
  var result = {
    presets: {}
  };

  var form = {
    pageid: 5,
    postif_registration_reject: 1,
    B15: set
  };

  if (!callback) {
    callback = set;

    talk ({ path: '/general/sleep.html' }, function (err, res) {
      if (err) {
        callback (err);
        return;
      }

      res.data.replace (/<select id="B15" name="B15" >(.+)<\/select>/, function (s, sel) {
        sel.replace (/<option value="(\d+)"( selected="selected")?>(\d+)&#32;Mins<\/option>/g, function (s2, a, b, c) {
          result.presets [parseInt (a, 10)] = parseInt (c, 10);

          if (b) {
            result.value = {
              key: parseInt (a, 10),
              minutes: parseInt (c, 10)
            };
          }
        });
      });

      callback (null, result);
    });

    return;
  }

  talk (
    {
      method: 'POST',
      path: '/general/sleep.html',
      query: form
    },
    function (err, res) {
      var error = null;

      if (err) {
        callback (err);
        return;
      }

      if (!!~res.data.indexOf ('<div class="postSuccess">')) {
        callback (null, true);
        return;
      }

      error = new Error ('Unknown error');
      error.statusCode = res.code;
      error.body = res.data;

      res.data.replace (/<div class="postError">([^<]+)<\/div>/, function (str, message) {
        error = new Error ('POST failed');
        error.statusCode = res.code;
        error.error = message.replace ('&#32;', ' ');
      });

      callback (error);
    }
  );
}


/**
 * Method: current
 *
 * @callback callback
 * @param callback {function} - `function (err, data) {}`
 * @returns {void}
 */

function methodCurrent (callback) {
  var printer = ipp.Printer ('http://' + config.hostname + ':' + config.ippPort + '/ipp/printer');
  var msg = {
    'operation-attributes-tag': {
      'requested-attributes': [
        'queued-job-count',
        'marker-names',
        'marker-levels',
        'printer-state',
        'printer-state-reasons',
        'printer-up-time'
      ]
    }
  };

  printer.execute ('Get-Printer-Attributes', msg, function (err, data) {
    var error = null;
    var result = {};
    var name;
    var i;

    if (err) {
      error = new Error ('ipp error');
      error.error = err;
      callback (err);
      return;
    }

    data = data ['printer-attributes-tag'] || {};

    result.state = data ['printer-state'] || null;
    result.stateReasons = data ['printer-state-reasons'] || null;
    result.jobs = data ['queued-job-count'] || 0;
    result.uptime = data ['printer-up-time'] || 0;
    result.uptimeDate = new Date (Date.now () - (result.uptime * 1000));

    result.ink = {};

    if (data ['marker-names'] instanceof Array && data ['marker-levels'] instanceof Array) {
      for (i = 0; i < data ['marker-names'] .length; i++) {
        name = data ['marker-names'] [i] .toLowerCase () .replace (/.*\u001e(\w+) .*/, '$1');
        result.ink [name] = data ['marker-levels'] [i];
      }
    }

    callback (null, result);
  });
}


/**
 * Module interface
 *
 * @param setup {object} - Configuration parameters
 * @param setup.hostname {string} - Printer hostname or IP
 * @param [setup.port = 80] {number} - Printer UI port
 * @param [setup.ippPort = 631] {number} - Printer IPP port
 * @param [setup.prefix] {string} - Path prefix, for proxies
 * @param [setup.protocol = http] {string} - Printer UI protocol, `http` or `https`
 * @returns {object} - Methods
 */

module.exports = function (setup) {
  var i;

  for (i in setup) {
    config [i] = setup [i];
  }

  return {
    current: methodCurrent,
    sleep: methodSleep,
    general: {
      status: methodGeneralStatus,
      information: methodGeneralInformation
    }
  };
};
